src/test/test_stats.c

```diff
@@ -246,6 +247,244 @@ test_rephist_mtbf(void *arg)
+    tt_str_op(buf, OP_EQ, str); \
+    tt_int_op(rv, OP_EQ, checkrv); \
+    tor_free(buf); \
+  STMT_END;
+
+static void
+test_fill_bandwidth_history(void *arg)
+{
+  (void) arg;
+  bw_array_t *b = bw_array_new();
+  char *buf;
+  size_t rv;
+  /* Remember bandwidth is rounded down to the nearest 1K. */
+  /* Day 1. */
+  set_test_case(b, 0, 0, 0, 0, 0, 0, 0);
+  buf = tor_malloc_zero(MAX_HIST_VALUE_LEN);
+  rv = rep_hist_fill_bandwidth_history(buf, MAX_HIST_VALUE_LEN, b);
+  tt_int_op(rv, OP_EQ, 0);
+  tor_free(buf);
+  /* Day 2. */
+  set_test_case(b, 1, 1, 1000, 0, 0, 0, 0);
+  test_fill_bw(b, buf, rv, "0", 1);
+  /* Day 3. */
+  set_test_case(b, 2, 2, 1000, 1500, 0, 0, 0);
+  test_fill_bw(b, buf, rv, "0,1024", 6);
+  /* Day 4. */
+  set_test_case(b, 3, 3, 1000, 1500, 3500, 0, 0);
+  test_fill_bw(b, buf, rv, "0,1024,3072", 11);
+  /* Day 5. */
+  set_test_case(b, 4, 4, 1000, 1500, 3500, 8000, 0);
+  test_fill_bw(b, buf, rv, "0,1024,3072,7168", 16);
+  /* Day 6. */
+  set_test_case(b, 5, 0, 1000, 1500, 3500, 8000, 6000);
+  test_fill_bw(b, buf, rv, "0,1024,3072,7168,5120", 21);
+  /* Day 7. */
+  /* Remember oldest entry first. */
+  set_test_case(b, 5, 1, 10000, 1500, 3500, 8000, 6000);
+  test_fill_bw(b, buf, rv, "1024,3072,7168,5120,9216", 24);
+  /* Mocking get_options to manipulate RelayBandwidthRate. */
+  MOCK(get_options, mock_get_options);
+  /* Limits bandwidth to 1 KBps. */
+  /* Cutoff is set to 88473600. */
+  mock_options.RelayBandwidthRate = 1024;
+  set_test_case(b, 5, 2, 88573600, 88473600, 10000, 8000, 6000);
+  test_fill_bw(b, buf, rv, "9216,7168,5120,88473600,88473600", 32);
+ done:
+  UNMOCK(get_options);
+  bw_array_free(b);
+}
+
+#define set_test_bw_lines(r, w, dr, dw, when) STMT_BEGIN \
+    rep_hist_note_bytes_read(r, when); \
+    rep_hist_note_bytes_written(w, when); \
+    rep_hist_note_dir_bytes_read(dr, when); \
+    rep_hist_note_dir_bytes_written(dw, when); \
+  STMT_END;
+
+#define test_get_bw_lines(str, checkstr) STMT_BEGIN \
+    str = rep_hist_get_bandwidth_lines(); \
+    tt_str_op(str, OP_EQ, checkstr); \
+    tor_free(str); \
+  STMT_END;
+
+static void
+test_get_bandwidth_lines(void *arg)
+{
+  (void) arg;
+  char *str, *checkstr;
+  char t[ISO_TIME_LEN+1];
+  int len = (67+MAX_HIST_VALUE_LEN)*4;
+  checkstr = tor_malloc_zero(len);
+  time_t now = time(NULL);
+  bw_arrays_init();
+
+  /* Day 1. */
+  now += 86400;
+  set_test_bw_lines(5000, 5500, 3000, 3500, now - 6*60*60);
+  /* Day 2. */
+  now += 86400;
+  set_test_bw_lines(50000, 55000, 30000, 35000, now - 6*60*60);
+  /* Day 3. */
+  now += 86400;
+  set_test_bw_lines(25000, 27500, 15000, 17500, now - 6*60*60);
+  /* Day 4. */
+  now += 86400;
+  set_test_bw_lines(90000, 76000, 60000, 45000, now - 6*60*60);
+  /* Day 5. */
+  now += 86400;
+  set_test_bw_lines(500, 55000, 30000, 35000, now - 6*60*60);
+  set_test_bw_lines(0, 0, 0, 0, now);
+  format_iso_time(t, now);
+  tor_snprintf(checkstr, len, "write-history %s (86400 s) "
```

Comment:
Pass the length of checkstr to snprintf(), not a calculated length.
```suggestion
  tor_snprintf(checkstr, sizeof(checkstr), "write-history %s (86400 s) "
```

Reply:
Oh right. Now I understand what's happening here.

It's slightly safer to use stack-allocated buffers for snprintf(), so that the actual size and snprintf size never get out of sync. But this code is also ok.